### PR TITLE
Introduce type-descriptor precedence support to parser

### LINKED ISSUE #24035: Add type-desc precedence support to parser
**Description:**
Precedence in decreasing order

```
distinct T
T[]
T1&T2
T1|T2
function(args) returns T
```
Spec issue [513](https://github.com/ballerina-platform/ballerina-spec/issues/513)

### LINKED ISSUE #30225: Issue with type descriptor precedence
**Description:**
According to the type descriptor precedence table in the spec, I expected `Foo & readonly|int|string[] & readonly` and `(readonly & Foo)|int|(string[] & readonly)` to be the same.

**Steps to reproduce:**
The following results in a compilation error.
```ballerina
public function main() {
    Foo & readonly|int|string[] & readonly x = recordIntersection({});
}

type Foo record {|
    int i = 1;
|};

function recordIntersection(Foo foo) returns (readonly & Foo)|int|(string[] & readonly) => foo.cloneReadOnly();
```

```cmd
$ bal run foo.bal 

Compiling source
        foo.bal
ERROR [foo.bal:(2:5,2:43)] invalid intersection type 'Foo & readonly|int|string[] & readonly': no intersection
```

**Affected Versions:**
slalpha5

### PR DESCRIPTION
**Pull request by lochana-chathura**
## Purpose
$subject.

Fixes #24035 
Fixes #30225 

## Approach
N/A

## Samples
```bal
type customType distinct Foo & readonly & int|Baz[] & Qux[a:b] & boolean|distinct float;
```
 will be parsed as,
```bal
type customType ( ( (((distinct Foo) & readonly) & int) | ((Baz[] & (Qux[a:b])) & boolean) ) | (distinct float) );
```

<details>
  <summary><b>Complex Sample</b>  (<i>click to expand</i>)</summary>

```bal
type customType readonly & int | Baz[] & Qux[mod1:app] & boolean | distinct float | 
function() returns Foo & distinct Bar | Baz & Qux[0] & 
function() returns int[] & distinct distinct Bar[] | string[];
```

 will be parsed as,
```bal
type customType ( ( ((readonly & int)| ( (Baz[] & (Qux[mod1:app])) & boolean )) | (distinct float) ) | 
(function() returns ( (Foo & (distinct Bar)) | 
(( Baz & (Qux[0]) ) & function() returns ( ( int[] & ( (distinct (distinct Bar))[] ) ) | (string[]) ) ) ) ) );
```
</details>

## Remarks
Procedence from high to low as follows.
```bal
distinct T
T[]
T1 & T2
T1 | T2
function(args) returns T
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by lochana-chathura** [comment]
Will fix the build failure. Parser changes are done.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -2322,26 +2336,79 @@ private STNode parseComplexTypeDescriptorInternal(STNode typeDesc, ParserRuleCon
...
+                newTypeDesc = parseIntersectionTypeDescriptor(typeDesc, context, isTypedBindingPattern);
+                return parseComplexTypeDescriptorInternal(newTypeDesc, context, isTypedBindingPattern, precedence);
             default:
                 return typeDesc;
         }
     }
 
-    private boolean isValidTypeContinuationToken(STToken nextToken) {
-        switch (nextToken.kind) {
+    private boolean isArrayTypeDescInTypedBP() {
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:2372]: What is the need of this method?

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java]: Addressed in 71a6f9e

**Comment by lochana-chathura** [comment]
@shehan360 docgen upated with 94e7804.

File: misc/docerina/src/main/java/org/ballerinalang/docgen/generator/model/Type.java [hunk 2]
```diff
@@ -276,6 +268,52 @@ public static Type fromNode(Node node, SemanticModel semanticModel) {
...
+        }
+
+        // flatten the types inside the union and add it to the list.
+        UnionTypeDescriptorNode unionType = (UnionTypeDescriptorNode) typeNode;
+        Node lhsType = unionType.leftTypeDesc();
+        if (lhsType.kind() != SyntaxKind.UNION_TYPE_DESC) {
+            memberTypes.add(fromNode(lhsType, semanticModel));
+        } else {
+            addUnionMemberTypes(memberTypes, lhsType, semanticModel);
+        }
```
Comment (rdulmina) [inline misc/docerina/src/main/java/org/ballerinalang/docgen/generator/model/Type.java:284]: We can extract this to a separate method and call with leftType and rightType

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java [hunk 3]
```diff
@@ -884,15 +884,32 @@ public BLangNode transform(UnionTypeDescriptorNode unionTypeDescriptorNode) {
...
+        } else {
+            flattenUnionType(list, lhsTypeDescNode);
+        }
+
+        TypeDescriptorNode rhsTypeDescNode = unionTypeDescriptorNode.rightTypeDesc();
+        if (rhsTypeDescNode.kind() != SyntaxKind.UNION_TYPE_DESC) {
+            list.add(rhsTypeDescNode);
+        } else {
+            flattenUnionType(list, rhsTypeDescNode);
+        }
```
Comment (rdulmina) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java:910]: lets extract out this as well

**Review by shehan360 — APPROVED** [review]

**Review by rdulmina — APPROVED** [review]

**Pull request merged**